File: surface/widgets/nativescript/components/DesignTokenProvider.ts
// ============================================================
// Clef Surface NativeScript Widget — DesignTokenProvider
//
// Provides resolved design tokens to descendant NativeScript
// components via an observable context pattern. Converts CSS
// token values to NativeScript-compatible color/dimension values.
// ============================================================

import { Observable, StackLayout, Label, EventData } from '@nativescript/core';

import type {
  DesignTokenValue,
  ResolvedTheme,
  Signal,
  ThemeConfig,
} from '../../shared/types.js';
import { resolveTheme } from '../../shared/surface-bridge.js';

// --------------- Context ---------------

export interface DesignTokenContextValue {
  theme: ResolvedTheme;
  tokens: Record<string, string>;
  getToken(name: string): string | undefined;
}

let _currentContext: DesignTokenContextValue | null = null;

export function getDesignTokens(): DesignTokenContextValue {
  if (!_currentContext) {
    throw new Error('getDesignTokens must be called within a DesignTokenProvider scope.');
  }
  return _currentContext;
}

// --------------- Props ---------------

export interface DesignTokenProviderProps {
  tokens?: DesignTokenValue[];
  themes?: ThemeConfig[];
  resolvedTheme?: ResolvedTheme;
  themeSignal?: Signal<ResolvedTheme>;
  showBorder?: boolean;
  label?: string;
}

// --------------- Component ---------------

export function createDesignTokenProvider(props: DesignTokenProviderProps = {}): StackLayout {
  const {
    tokens = [],
    themes = [],
    resolvedTheme,
    themeSignal,
    showBorder = false,
    label = 'tokens',
  } = props;

  const container = new StackLayout();
  container.className = 'clef-design-token-provider';

  let theme: ResolvedTheme;
  if (themeSignal) {
    theme = themeSignal.get();
  } else if (resolvedTheme) {
    theme = resolvedTheme;
  } else {
    theme = resolveTheme(tokens, themes);
  }

  _currentContext = {
    theme,
    tokens: theme.tokens,
    getToken: (name: string) => theme.tokens[name],
  };

  if (showBorder) {
    container.borderWidth = 1;
    container.borderColor = '#666666';
    container.padding = 8;

    const headerLabel = new Label();
    headerLabel.text = `[${label}: ${theme.name}]`;
    headerLabel.className = 'clef-token-header';
    headerLabel.opacity = 0.6;
    container.addChild(headerLabel);
  }

  if (themeSignal) {
    themeSignal.subscribe((newTheme) => {
      theme = newTheme;
      _currentContext = {
        theme: newTheme,
        tokens: newTheme.tokens,
        getToken: (name: string) => newTheme.tokens[name],
      };
    });
  }

  return container;
}

createDesignTokenProvider.displayName = 'DesignTokenProvider';
export default createDesignTokenProvider;

File: surface/widgets/compose/components/DesignTokenProvider.kt
// ============================================================
// Clef Surface Compose Widget — DesignTokenProvider
//
// CompositionLocal provider that supplies resolved design tokens
// to descendant Compose components. Provides token lookup via
// Compose's ambient state system (CompositionLocal).
// ============================================================

package clef.surface.compose.components

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

// --------------- Context ---------------

data class DesignTokenContextValue(
    val theme: ResolvedTheme,
    val tokens: Map<String, String>,
    val getToken: (String) -> String?,
)

val LocalDesignTokens = compositionLocalOf<DesignTokenContextValue?> { null }

// --------------- Hook ---------------

@Composable
fun rememberDesignTokens(): DesignTokenContextValue {
    return LocalDesignTokens.current
        ?: error("rememberDesignTokens must be used within a DesignTokenProvider.")
}

// --------------- Component ---------------

@Composable
fun DesignTokenProvider(
    tokens: List<DesignTokenValue> = emptyList(),
    themes: List<ThemeConfig> = emptyList(),
    resolvedTheme: ResolvedTheme? = null,
    showBorder: Boolean = false,
    label: String = "tokens",
    content: @Composable () -> Unit,
) {
    val theme = remember(resolvedTheme, tokens, themes) {
        resolvedTheme ?: resolveTheme(tokens, themes)
    }

    val contextValue = remember(theme) {
        DesignTokenContextValue(
            theme = theme,
            tokens = theme.tokens,
            getToken = { name -> theme.tokens[name] },
        )
    }

    CompositionLocalProvider(LocalDesignTokens provides contextValue) {
        if (showBorder) {
            Column(
                modifier = Modifier
                    .border(1.dp, MaterialTheme.colorScheme.outline)
                    .padding(horizontal = 8.dp)
            ) {
                Text(
                    text = "[$label: ${theme.name}]",
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    style = MaterialTheme.typography.labelSmall,
                )
                content()
            }
        } else {
            Column {
                content()
            }
        }
    }
}
